src/test/socket_utils.rs

```diff
@@ -0,0 +1,303 @@
+            assert_eq!(0, unsafe {
+                libc::socketpair(init_method.domain(), sock_type | flags, 0, fds.as_mut_ptr())
+            });
+
+            let (fd_client, fd_peer) = (fds[0], fds[1]);
+            assert!(fd_client >= 0 && fd_peer >= 0);
+
+            (fd_client, fd_peer)
+        }
+        SocketInitMethod::Inet | SocketInitMethod::Unix => {
+            let fd_client = unsafe { libc::socket(init_method.domain(), sock_type | flags, 0) };
+            let fd_server = unsafe { libc::socket(init_method.domain(), sock_type | flags, 0) };
+            assert!(fd_client >= 0);
+            assert!(fd_server >= 0);
+
+            // bind the server socket to some unused address
+            let (server_addr, server_addr_len) = autobind_helper(fd_server, init_method.domain());
+
+            if bind_client {
+                // bind the client socket to some unused address
+                autobind_helper(fd_client, init_method.domain());
+            }
+
+            match sock_type {
+                libc::SOCK_STREAM | libc::SOCK_SEQPACKET => {
+                    // connect the client to the server and get the accepted socket
+                    let fd_peer = stream_connect_helper(
+                        fd_client,
+                        fd_server,
+                        server_addr,
+                        server_addr_len,
+                        flags,
+                    );
+
+                    // close the server
+                    assert_eq!(0, unsafe { libc::close(fd_server) });
+
+                    (fd_client, fd_peer)
+                }
+                libc::SOCK_DGRAM => {
+                    // connect the client to the server
+                    dgram_connect_helper(fd_client, server_addr, server_addr_len);
+                    (fd_client, fd_server)
+                }
+                _ => unimplemented!(),
+            }
+        }
+    }
+}
+
+/// A helper function to connect the client socket to the server address. Returns the accepted
+/// socket.
+pub fn stream_connect_helper(
+    fd_client: libc::c_int,
+    fd_server: libc::c_int,
+    addr: SockAddr,
+    addr_len: libc::socklen_t,
+    flags: libc::c_int,
+) -> libc::c_int {
+    // listen for connections
+    {
+        let rv = unsafe { libc::listen(fd_server, 10) };
+        assert_eq!(rv, 0);
+    }
+
+    let addr_ptr = match addr {
+        SockAddr::Inet(ref x) => x as *const libc::sockaddr_in as *const libc::sockaddr,
+        SockAddr::Unix(ref x) => x as *const libc::sockaddr_un as *const libc::sockaddr,
+        _ => unimplemented!(),
+    };
+
+    // connect to the server address
+    {
+        let rv = unsafe { libc::connect(fd_client, addr_ptr, addr_len) };
+        assert!(rv == 0 || (rv == -1 && get_errno() == libc::EINPROGRESS));
+    }
+
+    // shadow needs to run events, otherwise the accept call won't know it
+    // has an incoming connection (SYN packet)
```

Comment:
If we didn't do this, wouldn't the `accept4` just block until the packet arrives? If the `usleep` is only needed for `SOCK_NONBLOCK`, maybe worth mentioning that explicitly. (And maybe only doing it if `SOCK_NONBLOCK` is set?)

Reply:
Yeah it's only needed for the non-blocking cases. I added a `if (flags & libc::SOCK_NONBLOCK) != 0` condition here.